server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -1000,30 +995,32 @@ private boolean tryRelocateShard(ModelNode minNode, ModelNode maxNode, String id
                 for (ShardRouting shard : index) {
                     if (shard.started()) {
                         // skip initializing, unassigned and relocating shards we can't relocate them anyway
-                        Decision allocationDecision = deciders.canAllocate(shard, minNode.getRoutingNode(), allocation);
-                        Decision rebalanceDecision = deciders.canRebalance(shard, allocation);
-                        if (((allocationDecision.type() == Type.YES) || (allocationDecision.type() == Type.THROTTLE))
-                                && ((rebalanceDecision.type() == Type.YES) || (rebalanceDecision.type() == Type.THROTTLE))) {
-                            if (maxNode.containsShard(shard)) {
-                                // simulate moving shard from maxNode to minNode
-                                final float delta = weight.weightShardAdded(
-                                    this, minNode, idx) - weight.weightShardRemoved(this, maxNode, idx);
-                                if (delta < minCost ||
-                                        (candidate != null && Float.compare(delta, minCost) == 0 && candidate.id() > shard.id())) {
-                                    /* this last line is a tie-breaker to make the shard allocation alg deterministic
-                                     * otherwise we rely on the iteration order of the index.getAllShards() which is a set.*/
-                                    minCost = delta;
-                                    candidate = shard;
-                                    decision = new Decision.Multi().add(allocationDecision).add(rebalanceDecision);
-                                }
-                            }
+                        if (maxNode.containsShard(shard)) {
+                            shardRoutings.add(shard);
                         }
                     }
                 }
 
+                // look for a relocation candidate, in descending order of shard id so that the decision is deterministic
+                shardRoutings.sort(BY_DESCENDING_SHARD_ID);
+                for (ShardRouting shard : shardRoutings) {
+                    final Decision rebalanceDecision = deciders.canRebalance(shard, allocation);
+                    if ((rebalanceDecision.type() != Type.YES) && (rebalanceDecision.type() != Type.THROTTLE)) {
+                        continue;
+                    }
+                    final Decision allocationDecision = deciders.canAllocate(shard, minNode.getRoutingNode(), allocation);
+                    if ((allocationDecision.type() != Type.YES) && (allocationDecision.type() != Type.THROTTLE)) {
+                        continue;
+                    }
+                    candidate = shard;
+                    decision = new Decision.Multi().add(allocationDecision).add(rebalanceDecision);
+                    break;
+                }
+
                 if (candidate != null) {
```

Comment:
Should we move this up as well, so that we can avoid the `if (candidate != null) {` check (as it's only assigned once now)?

Reply:
Yes, thanks, that's nicer still. ad503a5b0573eb9617c95c7403e133966edbe727.